optimum/neuron/modeling.py

```diff
@@ -653,3 +660,286 @@ def prepare_inputs_for_generation(self, input_ids: torch.Tensor, **kwargs) -> Di
+        else:
+            last_warper = logits_warper[-1]
+            fast_topk = isinstance(last_warper, TopKLogitsWarper)
+            if fast_topk:
+                # Replace the last warping operation by a faster alternative
+                logits_warper[-1] = self.FastTopKLogitsWarper(last_warper.top_k, last_warper.filter_value)
+
+        pad_token_id = pad_token_id if pad_token_id is not None else self.generation_config.pad_token_id
+        eos_token_id = eos_token_id if eos_token_id is not None else self.generation_config.eos_token_id
+        if isinstance(eos_token_id, int):
+            eos_token_id = [eos_token_id]
+        eos_token_id_tensor = torch.tensor(eos_token_id).to(input_ids.device) if eos_token_id is not None else None
+        output_scores = output_scores if output_scores is not None else self.generation_config.output_scores
+        output_attentions = (
+            output_attentions if output_attentions is not None else self.generation_config.output_attentions
+        )
+        output_hidden_states = (
+            output_hidden_states if output_hidden_states is not None else self.generation_config.output_hidden_states
+        )
+        return_dict_in_generate = (
+            return_dict_in_generate
+            if return_dict_in_generate is not None
+            else self.generation_config.return_dict_in_generate
+        )
+
+        # init attention / hidden states / scores tuples
+        scores = () if (return_dict_in_generate and output_scores) else None
+        decoder_attentions = () if (return_dict_in_generate and output_attentions) else None
+        cross_attentions = () if (return_dict_in_generate and output_attentions) else None
+        decoder_hidden_states = () if (return_dict_in_generate and output_hidden_states) else None
+
+        # keep track of which sequences are already finished
+        unfinished_sequences = torch.ones(input_ids.shape[0], dtype=torch.long, device=input_ids.device)
+
+        # auto-regressive generation
+        while True:
+            # prepare model inputs
+            model_inputs = self.prepare_inputs_for_generation(input_ids, **model_kwargs)
+
+            # forward pass to get next token
+            outputs = self(
+                **model_inputs,
+                return_dict=True,
+                output_attentions=output_attentions,
+                output_hidden_states=output_hidden_states,
+            )
+
+            next_token_logits = outputs.logits[:, -1, :]
+
+            # pre-process distribution
+            next_token_scores = logits_processor(input_ids, next_token_logits)
+            if fast_topk:
+                # Get [batch_size, top_k] scores and indices instead of [batch_size, vocab_size] scores
+                next_token_scores, next_token_indices = logits_warper(input_ids, next_token_scores)
+            else:
+                next_token_scores = logits_warper(input_ids, next_token_scores)
```

Comment:
That is the specific part we added? 
If not, let's comment to show where we differ from the original implementation.

Reply:
OK, I will do that. 

You can track the specific changes in the commits though: I first copied the method in one commit and then modified it in a second commit.
I understand that all commits in a pull-request are squashed, but in that particular configuration I find it convenient to keep separate commits.

Reply:
It is more to be able to know fast what is neuron-specific vs what comes from transformers.